File: cli/src/doctor.ts
import { constants } from "node:fs";
import { access, readdir, readFile } from "node:fs/promises";
import { join } from "node:path";
import { getAudioPlayerStatus } from "./audio.js";
import { loadConfig } from "./config.js";

type DoctorIssue = {
  title: string;
  detail: string;
  commands: string[];
};

type LinuxInputStatus = {
  hasInputDir: boolean;
  eventCount: number;
  readableCount: number;
  inputGroupExists: boolean;
  inputGroupActive: boolean;
  username: string;
};

export async function runDoctor() {
  const issues: DoctorIssue[] = [];

  console.log("Cliks doctor");
  console.log("");
  console.log("Privacy:");
  console.log("- Cliks sends only event kind: keyboard or mouse.");
  console.log("- Cliks sends timing offsets inside each 500ms batch.");
  console.log("- Cliks does not send key values, key codes, words, coordinates, windows, or app names.");
  console.log("");
  console.log("System:");

  const config = await loadConfig();
  const nodeMajor = Number(process.versions.node.split(".")[0]);
  if (nodeMajor >= 20) {
    console.log(`- Node.js: ok (${process.versions.node})`);
  } else {
    console.log(`- Node.js: needs update (${process.versions.node})`);
    issues.push({
      title: "Update Node.js",
      detail: "Cliks needs Node.js 20 or newer.",
      commands: ["node --version", "Install Node.js 20 or newer, then rerun: typ doctor"]
    });
  }

  const audio = getAudioPlayerStatus();
  if (audio.player) {
    console.log(`- Audio player: ok (${audio.player})`);
  } else {
    console.log("- Audio player: missing");
    issues.push({
      title: "Install an audio playback tool",
      detail: audio.hint ?? "Cliks could not find a supported audio player.",
      commands: [...audio.commands, "typ sound-test"]
    });
  }

  console.log(`- Platform: ${process.platform}`);
  console.log(`- Current team: ${config.currentTeamCode ?? "not joined"}`);
  console.log(`- Sharing keyboard: ${config.sharing.keyboard ? "on" : "off"}`);
  console.log(`- Sharing mouse: ${config.sharing.mouse ? "on" : "off"}`);

  if (!config.currentTeamCode) {
    issues.push({
      title: "Join a team",
      detail: "Cliks does not have a selected team code.",
      commands: ["typ join CLIK-XXXX"]
    });
  }

  if (!config.sharing.keyboard) {
    issues.push({
      title: "Turn keyboard sharing on",
      detail: "Your keystroke timing will not reach teammates while share.keyboard is off.",
      commands: ["typ set share.keyboard on"]
    });
  }

  if (!config.sharing.mouse) {
    issues.push({
      title: "Turn mouse sharing on",
      detail: "Your mouse click timing will not reach teammates while share.mouse is off.",
      commands: ["typ set share.mouse on"]
    });
  }

  if (process.platform === "linux") {
    const input = await getLinuxInputStatus();
    console.log(`- Linux input devices: ${input.hasInputDir ? "found" : "not found"}`);
    if (input.hasInputDir) {
      console.log(`- Readable event devices: ${input.readableCount}/${input.eventCount}`);
      console.log(`- Active input group: ${input.inputGroupActive ? "yes" : "no"}`);
    }

    if (!input.hasInputDir) {
      issues.push({
        title: "Global capture is unavailable in this environment",
        detail: "Cliks cannot see /dev/input. This is normal in Termux, containers, SSH sessions, and some locked-down environments.",
        commands: ["Use a normal desktop terminal for global capture", "For a local terminal-only test: typ start --terminal --self"]
      });
    } else if (input.eventCount === 0) {
      issues.push({
        title: "No input event devices found",
        detail: "Cliks found /dev/input, but there are no /dev/input/event* devices to read.",
        commands: ["ls -l /dev/input", "Try again from the real desktop session"]
      });
    } else if (input.readableCount === 0) {
      const commands = input.inputGroupExists
        ? [`sudo usermod -aG input ${input.username}`, "Log out and log back in, or reboot", "typ doctor"]
        : ["Ask your distro how to allow reading /dev/input/event* for global input capture", "typ doctor"];
      issues.push({
        title: "Allow Cliks to read input events",
        detail: "Linux global capture needs permission to read /dev/input/event*. Cliks still sends only event type and timing, never key values.",
        commands
      });
    } else if (!input.inputGroupActive && input.inputGroupExists) {
      issues.push({
        title: "Input permission may not survive every device",
        detail: "Some event devices are readable, but your current login session is not in the input group.",
        commands: [`sudo usermod -aG input ${input.username}`, "Log out and log back in, or reboot", "typ doctor"]
      });
    }

    console.log("");
    console.log("Recommended run command:");
    console.log(input.readableCount > 0 ? "  typ start --evdev" : "  typ start --terminal --self");
  } else if (process.platform === "darwin") {
    issues.push({
      title: "Allow Accessibility permission",
      detail: "macOS global input capture needs Accessibility permission for the terminal app.",
      commands: ["Open System Settings > Privacy & Security > Accessibility", "Allow your terminal app", "typ doctor"]
    });
  } else if (process.platform === "win32") {
    console.log("");
    console.log("Recommended run command:");
    console.log("  typ start");
  } else {
    issues.push({
      title: "Unsupported global capture platform",
      detail: "This platform is not fully supported for global input capture yet.",
      commands: ["typ start --terminal --self"]
    });
  }

  console.log("");
  if (issues.length === 0) {
    console.log("No blocking issues detected.");
    console.log("Test playback: typ sound-test");
    return;
  }

  console.log("Fixes:");
  for (const issue of issues) {
    console.log("");
    console.log(`${issue.title}:`);
    console.log(`  ${issue.detail}`);
    for (const command of issue.commands) {
      console.log(`  ${command}`);
    }
  }
}

async function getLinuxInputStatus(): Promise<LinuxInputStatus> {
  const username = process.env.USER || process.env.LOGNAME || "$USER";
  const groupInfo = await readInputGroup();
  const activeGroups = typeof process.getgroups === "function" ? process.getgroups() : [];
  const inputGroupActive = groupInfo?.gid !== undefined && activeGroups.includes(groupInfo.gid);

  let entries: string[];
  try {
    entries = await readdir("/dev/input");
  } catch {
    return {
      hasInputDir: false,
      eventCount: 0,
      readableCount: 0,
      inputGroupExists: Boolean(groupInfo),
      inputGroupActive,
      username
    };
  }

  const eventDevices = entries.filter((entry) => /^event\d+$/.test(entry)).map((entry) => join("/dev/input", entry));
  let readableCount = 0;
  for (const device of eventDevices) {
    try {
      await access(device, constants.R_OK);
      readableCount += 1;
    } catch {
      // Count unreadable devices without failing the whole doctor run.
    }
  }

  return {
    hasInputDir: true,
    eventCount: eventDevices.length,
    readableCount,
    inputGroupExists: Boolean(groupInfo),
    inputGroupActive,
    username
  };
}

async function readInputGroup() {
  try {
    const groups = await readFile("/etc/group", "utf8");
    const line = groups.split("\n").find((entry) => entry.startsWith("input:"));
    if (!line) return undefined;
    const gid = Number(line.split(":")[2]);
    return Number.isFinite(gid) ? { gid } : undefined;
  } catch {
    return undefined;
  }
}

File: cli/doctor.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

type doctorIssue struct {
	title    string
	detail   string
	commands []string
}

type doctorCheck struct {
	label  string
	status string
}

type doctorReport struct {
	privacy        []string
	checks         []doctorCheck
	recommendation []string
	issues         []doctorIssue
}

func buildDoctorReport(cfg CliksConfig) doctorReport {
	return buildDoctorReportOptions(cfg, true)
}

// buildDoctorReportOptions builds diagnostics. When thorough is false, skip probes that
// start global hooks so passive startup notices stay instant.
func buildDoctorReportOptions(cfg CliksConfig, thorough bool) doctorReport {
	report := doctorReport{privacy: []string{
		"Cliks sends only event kind: keyboard or mouse.",
		"Cliks relays coarse timing offsets inside each 500ms batch.",
		"Cliks does not send key values, key codes, words, coordinates, windows, or app names.",
	}}
	report.checks = append(report.checks, doctorCheck{"Runtime", "Go " + runtime.Version()})

	player, spatial, hint, commands := getAudioPlayerStatus(cfg.Listening.AudioDevice)
	if player != "" {
		report.checks = append(report.checks, doctorCheck{"Audio player", "ok (" + player + ")"})
		if cfg.Listening.AudioDevice != "" {
			if hint == "" {
				report.checks = append(report.checks, doctorCheck{"Audio output", cfg.Listening.AudioDevice})
			} else {
				report.issues = append(report.issues, doctorIssue{"Choose a supported audio output", hint, []string{"cliks set audio.device default"}})
			}
		}
		if spatial {
			report.checks = append(report.checks, doctorCheck{"Spatial audio", "stereo pan + distance"})
		} else {
			report.checks = append(report.checks, doctorCheck{"Spatial audio", "distance only — run cliks setup (installs mpv for stereo pan)"})
		}
	} else {
		report.checks = append(report.checks,
			doctorCheck{"Audio player", "missing"},
			doctorCheck{"Spatial audio", "unavailable until an audio player is installed"},
		)
		report.issues = append(report.issues, doctorIssue{"Install an audio playback tool", hint, append(commands, "cliks sound-test")})
	}

	report.checks = append(report.checks,
		doctorCheck{"Platform", runtime.GOOS},
		doctorCheck{"Current team", valuePlain(cfg.CurrentTeamCode, "not joined")},
		doctorCheck{"Sharing keyboard", yesNo(cfg.Sharing.Keyboard)},
		doctorCheck{"Sharing mouse", yesNo(cfg.Sharing.Mouse)},
	)
	if cfg.CurrentTeamCode == "" {
		report.issues = append(report.issues, doctorIssue{"Join a team", "Cliks does not have a selected team code.", []string{"cliks join CLIK-XXXXXX"}})
	}
	if !cfg.Sharing.Keyboard {
		report.issues = append(report.issues, doctorIssue{"Turn keyboard sharing on", "Your keyboard activity will not reach teammates while share.keyboard is off.", []string{"cliks set share.keyboard on"}})
	}
	if !cfg.Sharing.Mouse {
		report.issues = append(report.issues, doctorIssue{"Turn mouse sharing on", "Your mouse activity will not reach teammates while share.mouse is off.", []string{"cliks set share.mouse on"}})
	}

	appendPlatformCaptureChecks(&report, thorough)
	if warning := lastConfigLoadWarning(); warning != "" {
		report.issues = append([]doctorIssue{{
			title:    "Fix or reset saved settings",
			detail:   warning,
			commands: []string{"Open Preferences in Cliks and press s to save", "or delete the broken config file and run cliks setup"},
		}}, report.issues...)
		report.checks = append(report.checks, doctorCheck{"Config file", "invalid JSON — using defaults"})
	} else {
		report.checks = append(report.checks, doctorCheck{"Config file", configPath()})
	}
	return report
}

func doctorReportLines(report doctorReport) []string {
	lines := []string{"Cliks doctor", "", "Privacy:"}
	for _, item := range report.privacy {
		lines = append(lines, "- "+item)
	}
	lines = append(lines, "", "System:")
	for _, check := range report.checks {
		lines = append(lines, fmt.Sprintf("- %s: %s", check.label, check.status))
	}
	if len(report.recommendation) > 0 {
		lines = append(lines, "", report.recommendation[0])
		for _, command := range report.recommendation[1:] {
			lines = append(lines, "  "+command)
		}
	}
	lines = append(lines, "")
	if len(report.issues) == 0 {
		return append(lines, "All good. Optional: cliks sound-test", "If something ever feels off: cliks setup")
	}
	lines = append(lines, "Easy fixes (or run: cliks setup):")
	for _, issue := range report.issues {
		lines = append(lines, "", issue.title+":", "  "+issue.detail)
		for _, command := range issue.commands {
			lines = append(lines, "  "+command)
		}
	}
	return lines
}

func doctorSummary(report doctorReport) string {
	if len(report.issues) == 0 {
		return "All good. Sound and capture look ready."
	}
	if len(report.issues) == 1 {
		return "1 easy setup step left. Prefer: cliks setup"
	}
	return fmt.Sprintf("%d easy setup steps left. Prefer: cliks setup", len(report.issues))
}

func passiveDoctorWarning(report doctorReport) string {
	var titles []string
	for _, issue := range report.issues {
		if issue.title == "Join a team" || issue.title == "Turn keyboard sharing on" || issue.title == "Turn mouse sharing on" {
			continue
		}
		titles = append(titles, issue.title)
		if len(titles) == 2 {
			break
		}
	}
	if len(titles) == 0 {
		return ""
	}
	return "Setup note: " + strings.Join(titles, "; ") + ". Open Diagnostics in Cliks or run cliks doctor."
}

func runDoctor() error {
	fmt.Println(strings.Join(doctorReportLines(buildDoctorReport(loadConfig())), "\n"))
	return nil
}

func quickDoctorWarning(cfg CliksConfig) string {
	return passiveDoctorWarning(buildDoctorReportOptions(cfg, false))
}

type inputStatus struct {
	hasInputDir      bool
	eventCount       int
	readableCount    int
	inputGroupActive bool
	username         string
}

func linuxInputStatus() inputStatus {
	username := getenvDefault("USER", getenvDefault("LOGNAME", "$USER"))
	status := inputStatus{username: username}
	entries, err := os.ReadDir("/dev/input")
	if err != nil {
		return status
	}
	status.hasInputDir = true
	for _, entry := range entries {
		if !strings.HasPrefix(entry.Name(), "event") {
			continue
		}
		status.eventCount++
		if file, err := os.Open(filepath.Join("/dev/input", entry.Name())); err == nil {
			status.readableCount++
			_ = file.Close()
		}
	}
	groupBytes, err := os.ReadFile("/etc/group")
	if err == nil {
		for _, line := range strings.Split(string(groupBytes), "\n") {
			if strings.HasPrefix(line, "input:") {
				parts := strings.Split(line, ":")
				if len(parts) > 2 {
					gid, _ := strconv.Atoi(parts[2])
					groups, _ := os.Getgroups()
					for _, group := range groups {
						if group == gid {
							status.inputGroupActive = true
						}
					}
				}
			}
		}
	}
	return status
}

func runSoundTest() error {
	audio := newAudioEngine(loadConfig().Listening)
	defer audio.Close()
	if err := audio.preview(); err != nil {
		return err
	}
	fmt.Println("Played keyboard, keyboard, mouse, mouse test clicks.")
	return nil
}

func yesNo(value bool) string {
	if value {
		return "yes"
	}
	return "no"
}

func valuePlain(value, fallback string) string {
	if strings.TrimSpace(value) == "" {
		return fallback
	}
	return value
}

func hasCommand(command string) bool {
	_, err := exec.LookPath(command)
	return err == nil
}
